core/trino-main/src/main/java/io/trino/execution/buffer/SpoolingExchangeOutputBuffer.java

```diff
@@ -166,34 +167,51 @@ public void enqueue(int partition, List<Slice> pages)
     @Override
     public void setNoMorePages()
     {
-        stateMachine.noMorePages();
-        destroy();
+        if (stateMachine.noMorePages()) {
+            exchangeSink.finish().whenComplete((value, failure) -> {
+                try {
+                    if (failure != null) {
+                        stateMachine.fail(failure);
+                    }
+                    else {
+                        stateMachine.finish();
+                    }
+                }
+                finally {
+                    updateMemoryUsage(0);
+                }
+            });
+        }
     }
 
     @Override
     public void destroy()
     {
-        if (stateMachine.finish()) {
-            try {
-                exchangeSink.finish();
-            }
-            finally {
-                updateMemoryUsage(exchangeSink.getMemoryUsage());
-            }
+        if (stateMachine.getState().canAddPages()) {
```

Comment:
`setNoMorePages` starts running `finish` after transitioning the state. If `destroy` is called before `setNoMorePages` it means that the task got cancelled prematurely and the buffer has to be invalidated. If there's a race and `setNoMorePages` is called at the same time when `destroy` is called it is legit to `finish` the sink, as the data written to the sink at that point is complete.

Reply:
I don't fully understand. If say this race is fine, then why do we need `if (stateMachine.getState().canAddPages()) {` check here? In case of a race, if would be like this check does not exist.

Reply:
In normal flow there shouldn't be a race. When the output is completely written and the `setNoMorePages` is called the task is only finished after `ExchangeSink#finish` is done and the buffer is transitioned to the `FINISHED` state. When the task itself is transitioned to `FINISHED` the `destroy` method is called and we don't want the sink to be aborted under normal circumstances. That's why there's a check.

However a race is possible when all the data is written but the task is cancelled before `ExchangeSink#finish` is completed. This shouldn't happen in practice, as the scheduler is not expected to cancel tasks that are writing to a spooling exchange. However from the interface perspective it is possible. I was thinking about what's the best way to handle this situation. When the output is complete and the task is cancelled the output itself is valid. So letting it finish should be perfectly fine. However sending an "abort" to the sink gives the `ExchangeSink` implementation to cancel commit if possible.

Reply:
Discussed offline.

Removing the check to ensure `abort` is always called if the `finish` hasn't succeeded.